### DISCUSSION TITLE
Flow Control in district heating network

### DISCUSSION THREAD

**Discussion by PSeros** [e2nIEE/pandapipes#627] [Q&A] [answered]
I am currently in the process of modeling a district heating network.
I created every node and every edge for the flow and with x+2 for the return.
Every heat exchanger (house), connects the flow line with the return line.
And the implemented circ_pump_const_mass_flow is doing it reversed, with a set temperature and massflow
I particularly wanted to take into account the discussion at the following link.
(https://github.com/e2nIEE/pandapipes/issues/309)
The calculation of the pipe flow shown in the picture below works (both hydraulic and thermal) only if I set control_active=False for the flow controls.

Does anyone have an idea what it could be?

![Netzmodell](https://github.com/e2nIEE/pandapipes/assets/141262779/23ce7c96-9d10-420b-829b-fdd9e6a0222f)

**Comment by dlohmeier** [comment, marked as answer]
Hi @Phantom19213 ,
great to see pandapipes in action in a real grid model. Without knowing your code exactly, I assume that you ran into the issue of too strict boundary conditions. If you define flow controllers that are fed by a circulation pump with fixed mass, computational inaccuracies can lead to problems in finding a solution. I used to address this problem by creating a "bypass" or "shortcircuit" valve with extremely high resistance (i.e. loss coefficient), but by now I came to the conclusion that the circulation pump with constant mass flow is not a good model for district heating grids with pre-defined mass flows for consumers. (By the way, have you checked out the new heat consumer model? It should be  simpler to use than the combination of flow control and heat exchanger.) I would recommend to use a circulation pump with constant pressure lift and just think about what pressure levels make sense. It is actually quite simple to add a control loop for setting the pressure levels such that e.g. a minimal pressure drop at the "last" consumer is guaranteed.
Does that help you?
Kind regards!

**Reply by PSeros** [reply]
Thank you very much, the pump with const pressure works.
But i haven't worked with control loops in pandapipes yet. Do you perhaps have an example?

**Reply by EPrade** [reply]
@Phantom19213 
I think some kind of simplistic while-loop iteration could suffice in this case. The condition could be the minimal pressure drop at the "last" consumer. While that condition is not reached the pressure level of the circ pump is adjusted iteratively until the desired result is reached.

**Reply by PSeros** [reply]
@EPrade 
The only problem with a while loop would be that I edit everything in a Jupyter notebook in a couple of seperated code segments.
I was thinking more about the controller class implemented in pandapower. I've just never worked with them before and have no idea how to build them.

**Reply by JonasPfeiffer123** [reply]
I'd like to add that controlling the pressure drop at the "worst" consumer in the network works pretty good for me with a simple implementation of a pressure controller for the const_pressure pump. Basically increasing the pressure lift when the dp is below a defined min value.

**Reply by mathys-heuang** [reply]
@dlohmeier Hi! What do you mean by "the new heat consumer model" ? Is there another way than flow control + heat exchanger to model heat consumers in a district heating network ?

**Reply by JonasPfeiffer123** [reply]
@mathys-heuang they recently implemented a heat consumer component which combines flow control and heat exchanger in Version 0.10.0.

**Reply by dlohmeier** [reply]
Yes, that's the one I mean. On the develop branch, we have extended this heat consumer model to be able to control the mass flow to automatically meet a given return temperature. This requires a new calculation mode that we called bi-directional coupling, as we integrate hydraulic and heat transfer calculation to converge in one simulation run. (see PR #622 )

**Reply by JonasPfeiffer123** [reply]
@dlohmeier Oh interesting, haven't looked into that so far. I've implemented a custom controller for the return temperature a while back for my applications when using flow controls and heat exchangers and it kept working with the new heat consumer component. But this new approach sounds like a much faster solution.
